tests/bdd/test_volume_topology.py

```diff
@@ -0,0 +1,536 @@
+            common.get_pools_api().get_pool(POOL_2_UUID),
+        )
+        == 0
+    )
+
+
+@then("pool labels must not contain the volume request topology labels")
+def pool_labels_must_not_contain_the_volume_request_topology_labels(create_request):
+    """pool labels must not contain the volume request topology labels."""
+    assert common_labels(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ],
+        common.get_pools_api().get_pool(POOL_1_UUID),
+    ) or common_labels(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ],
+        common.get_pools_api().get_pool(POOL_2_UUID),
+    ) != len(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ]
+    )
+
+
+@then("replica addition should fail with an insufficient storage error")
+def replica_addition_should_fail_with_an_insufficient_storage_error(replica_ctx):
+    """replica addition should fail with an insufficient storage error."""
+    assert replica_ctx[REPLICA_ERROR] is not None
+    exception_info = replica_ctx[REPLICA_ERROR].__dict__
+    assert exception_info["status"] == requests.codes["insufficient_storage"]
+
+
+@then("volume creation should fail with an insufficient storage error")
+def volume_creation_should_fail_with_an_insufficient_storage_error(create_request):
+    """volume creation should fail with an insufficient storage error."""
+    request = create_request[CREATE_REQUEST_KEY]
+    try:
+        common.get_volumes_api().put_volume(VOLUME_UUID, request)
+    except Exception as e:
+        exception_info = e.__dict__
+        assert exception_info["status"] == requests.codes["insufficient_storage"]
+
+    # Check that the volume wasn't created.
+    volumes = common.get_volumes_api().get_volumes()
+    assert len(volumes) == 0
+
+
+@then("volume creation should succeed with a returned volume object with topology")
+def volume_creation_should_succeed_with_a_returned_volume_object_with_topology(
+    create_request,
+):
+    """volume creation should succeed with a returned volume object with topology."""
+    cfg = common.get_cfg()
+    expected_spec = VolumeSpec(
+        1,
+        VOLUME_SIZE,
+        SpecStatus("Created"),
+        VOLUME_UUID,
+        VolumePolicy(False),
+        topology={
+            "pool_topology": {
+                "labelled": {
+                    "inclusion": {"openebs.io/created-by": "msp-operator"},
+                    "exclusion": {},
+                }
+            }
+        },
+        _configuration=cfg,
+    )
+
+    # Check the volume object returned is as expected
+    request = create_request[CREATE_REQUEST_KEY]
+    volume = common.get_volumes_api().put_volume(VOLUME_UUID, request)
+    assert str(volume.spec) == str(expected_spec)
```

Comment:
Because this function is testing that volume creation should succeed, it might be good to also add checks for the volume state i.e. is the volume online